src/components/GlossaryList/GlossaryList.hooks.tsx

```diff
@@ -0,0 +1,126 @@
+          filteredGlossaryEntries.push(glossaryEntry)
+        }
+      })
+
+      return filteredGlossaryEntries
+    },
+    []
+  )
+
+  const onFilterByIndexElement = useCallback(
+    (glossaryEntries: GlossaryEntryProps[], selectedIndexElement?: string): GlossaryEntryProps[] => {
+      const filteredGlossaryEntries: GlossaryEntryProps[] = []
+
+      if (selectedIndexElement === null || selectedIndexElement === '' || selectedIndexElement === undefined) {
+        return glossaryEntries
+      }
+
+      if (selectedIndexElement === t('pages.glossary.fundamentals')) {
+        Array.from(glossaryEntries).forEach((glossaryEntry) => {
+          if (glossaryEntry.fundamental) {
+            filteredGlossaryEntries.push(glossaryEntry)
+          }
+        })
+
+        return filteredGlossaryEntries
+      }
+
+      Array.from(glossaryEntries).forEach((glossaryEntry) => {
+        if (
+          glossaryEntry.term &&
+          Array.from(glossaryEntry.term)[0].toLowerCase() === selectedIndexElement.toLowerCase()
+        ) {
+          filteredGlossaryEntries.push(glossaryEntry)
+        }
+      })
+
+      return filteredGlossaryEntries
+    },
+    [t]
+  )
+
+  const onSearchByQuery = useCallback(
+    (glossaryEntries: GlossaryEntryProps[], searchQuery?: string): GlossaryEntryProps[] => {
+      const searchedGlossaryEntries: GlossaryEntryProps[] = []
+
+      if (searchQuery === undefined || searchQuery === '') {
+        return glossaryEntries
+      }
+
+      Array.from(glossaryEntries).forEach((glossaryEntry) => {
+        if (
+          glossaryEntry.term?.toLowerCase().includes(searchQuery.toLowerCase()) ||
+          glossaryEntry.definition?.toLowerCase().includes(searchQuery.toLowerCase()) ||
+          glossaryEntry.sources?.toLowerCase().includes(searchQuery.toLowerCase()) ||
+          glossaryEntry.tags?.some((tag) => tag.toLowerCase().includes(searchQuery.toLowerCase()))
+        ) {
+          searchedGlossaryEntries.push(glossaryEntry)
+        }
+      })
+
+      return searchedGlossaryEntries
+    },
+    []
+  )
+
+  const onCollapseAll = useCallback((setExpandedList: (newExpandedList: string[]) => void) => {
+    setExpandedList?.([])
+  }, [])
+
+  const onExpandAll = useCallback(
+    (setExpandedList: (newExpandedList: string[]) => void, glossaryEntries: GlossaryEntryProps[]) => {
+      const tempExpandedList: string[] = []
+
+      Array.from(glossaryEntries).forEach((glossaryEntry) => {
+        glossaryEntry.term && tempExpandedList.push(glossaryEntry.term)
+      })
+
+      setExpandedList?.(tempExpandedList)
+    },
+    []
+  )
+
+  return useMemo(
+    () =>
+      ({
+        filterByTags: onFilterByTags,
+        filterByIndexElement: onFilterByIndexElement,
+        searchByQuery: onSearchByQuery,
+        collapseAll: onCollapseAll,
+        expandAll: onExpandAll
+      } as const),
+    [onFilterByTags, onFilterByIndexElement, onSearchByQuery, onCollapseAll, onExpandAll]
+  )
+}
```

Comment:
Kleiner refactor, damits kein erstellen von objekten bei Array.map gibt und das naming an die eigentliche Liste angepasst ist.
```suggestion
import { useCallback, useMemo } from 'react'
import { GlossaryEntryProps } from '@components'
import { useTranslation } from 'react-i18next'

export type GlossaryListHookReturn = {
  filterByTags: (glossaryEntries: GlossaryEntryProps[], selectedTags?: string[]) => GlossaryEntryProps[]
  filterByIndexElement: (glossaryEntries: GlossaryEntryProps[], selectedIndexElement?: string) => GlossaryEntryProps[]
  searchByQuery: (glossaryEntries: GlossaryEntryProps[], searchQuery?: string) => GlossaryEntryProps[]
  collapseAll: (setExpandedList?: (newExpandedList: string[]) => void) => void
  expandAll: (setExpandedList?: (newExpandedList: string[]) => void, glossaryEntries?: GlossaryEntryProps[]) => void
}

export const useGlossaryList = (): GlossaryListHookReturn => {
  const { t } = useTranslation()

  const filterByTags = useCallback(
    (glossaryEntries: GlossaryEntryProps[], selectedTags?: string[]): GlossaryEntryProps[] => {
      if (!selectedTags || selectedTags.length === 0) {
        return glossaryEntries
      }

      return glossaryEntries.filter((glossaryEntry) =>
        selectedTags.every((selectedTag) => glossaryEntry.tags?.includes(selectedTag))
      )
    },
    []
  )

  const filterByIndexElement = useCallback(
    (glossaryEntries: GlossaryEntryProps[], selectedIndexElement?: string): GlossaryEntryProps[] => {
      if (!selectedIndexElement)
        return glossaryEntries

      if (selectedIndexElement === t('pages.glossary.fundamentals'))
        return glossaryEntries.filter((glossaryEntry) => glossaryEntry.fundamental)

      return glossaryEntries.filter(
        (glossaryEntry) =>
          glossaryEntry.term?.toLowerCase().startsWith(selectedIndexElement.toLowerCase()) ?? false
      )
    },
    [t]
  )

  const searchByQuery = useCallback(
    (glossaryEntries: GlossaryEntryProps[], searchQuery?: string): GlossaryEntryProps[] => {
      if (!searchQuery) {
        return glossaryEntries
      }

      return glossaryEntries.filter(
        (glossaryEntry) =>
          glossaryEntry.term?.toLowerCase().includes(searchQuery.toLowerCase()) ??
          glossaryEntry.definition?.toLowerCase().includes(searchQuery.toLowerCase()) ??
          glossaryEntry.sources?.toLowerCase().includes(searchQuery.toLowerCase()) ??
          glossaryEntry.tags?.some((tag) => tag.toLowerCase().includes(searchQuery.toLowerCase())) ??
          false
      )
    },
    []
  )

  const collapseAll = useCallback((setExpandedList?: (newExpandedList: string[]) => void) => {
    setExpandedList?.([])
  }, [])

  const expandAll = useCallback(
    (setExpandedList?: (newExpandedList: string[]) => void, glossaryEntries?: GlossaryEntryProps[]) => {
      const tempExpandedList = glossaryEntries?.map(({ term }) => term).filter((term): term is string => term !== undefined) ?? []

      setExpandedList?.(tempExpandedList.filter((term): term is string => term !== undefined))
    },
    []
  )

  return useMemo(
    () => ({
      filterByTags,
      filterByIndexElement,
      searchByQuery,
      collapseAll,
      expandAll,
    }),
    [filterByTags, filterByIndexElement, searchByQuery, collapseAll, expandAll]
  )
}

```

Reply:
Ab Zeile 51: In Produktion funktioniert der vorgeschlagene Code. Beim Testen schlägt er allerdings fehlt mit den Nullish Coalescing Operators. Mit dem logischen Oder hingegen keine Probleme im Test. Jedoch meckert dort Sonarlint.